Port FV3-GSDCHEM-WW3 for GEFSv12 to WCOSS2

XianwuXue-NOAA opened the pull request.

RatkoVasic-NOAA committed:
Change in WCOSS account number from FV3GFS-T2O to GFS-DEV

WalterKolczynski-NOAA committed:
Merge pull request #1 from NOAA-EMC/release/gefs_v12

Release/gefs v12

XianwuXue-NOAA committed:
Change submodule pointer for FV3 from vlab to github

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules
	modified:   FV3
Refs: #348

WalterKolczynski-NOAA committed:
Update remaining submodules to GitHub

The repos for FV3 and GSDCHEM are updated to point to the repos on
GitHub.

Refs: #2

WalterKolczynski-NOAA committed:
Update FV3 model commit

Update FV3 to the current tip of production/GEFS_v12.

Refs: #2

WalterKolczynski-NOAA committed:
Merge pull request #4 from WalterKolczynski-NOAA/feature/migrate_github

Update remaining submodules to GitHub

XianwuXue-NOAA committed:
Update all submodules to Github

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules
	modified:   FV3

XianwuXue-NOAA committed:
Merge branch develop into feature/gefs_v12_github

XianwuXue-NOAA committed:
Add files to port to Acorn/WCOSS2

 On branch feature/gefs_v12_github
 Changes to be committed:
	new file:   conf/configure.fv3.wcoss2
	new file:   modulefiles/wcoss2/fv3

Refs: #3

XianwuXue-NOAA committed:
Update NEMS and WW3 commit ID

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   NEMS
	modified:   WW3

Refs: #3

XianwuXue-NOAA committed:
Optimize config files and update esmf version

1) Optimize conf/configure.fv3.wcoss2 for WCOSS2
2) Update esmf version to build it successfully

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   conf/configure.fv3.wcoss2
	modified:   modulefiles/wcoss2/fv3

Refs: #3

XianwuXue-NOAA committed:
Update FV3 repo from EMC to Xianwu.Xue

To use the FV3 ported WCOSS2 by fixing some incompatible
	codes

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules
	modified:   FV3
	modified:   modulefiles/wcoss2/fv3

Refs: #3

XianwuXue-NOAA committed:
Update esmf version to 8.0.1

Reasons:
1) snapshot26 is NOT available on Cactus
2) dell_p35 (prod) are using ESMF/8.0.0

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   modulefiles/wcoss2/fv3

Refs: #3

XianwuXue-NOAA committed:
Update WW3 commit id

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   WW3

Refs: #3

XianwuXue-NOAA committed:
Add -fast-transcendentals

Add "-fast-transcendentals" based on Recommended Intel compiler flags to achieve best
    performance based on the WCOSS2 wiki
 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   conf/configure.fv3.wcoss2

Refs: #23

XianwuXue-NOAA committed:
Update GSDCHEM to use the ported version

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules
	modified:   GSDCHEM

Refs: #23

XianwuXue-NOAA committed:
Update FV3 commit id

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   FV3

Refs: #23

XianwuXue-NOAA committed:
Change submodule pointer for FV3 back to EMC

Update the FV3 commit to the commit ID in the merged branch in EMC

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules
	modified:   FV3

Refs: #23

XianwuXue-NOAA committed:
Update some modules using the new versions

 On branch feature/gefs_v12_github
 Your branch is up to date with 'origin/feature/gefs_v12_github'.
	modified:   modulefiles/wcoss2/fv3

Refs: #23

XianwuXue-NOAA committed:
Update NEMS & build module to remove hardwired ver

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   NEMS
	modified:   modulefiles/wcoss2/fv3

Refs: #23

XianwuXue-NOAA committed:
Reomve dependency on sfcio lib

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   modulefiles/wcoss2/fv3

Refs: #23

XianwuXue-NOAA committed:
Change HDF5 to HDF5_ROOT since HDF5 doesn't exist

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   conf/configure.fv3.wcoss2

Refs: #23

XianwuXue-NOAA committed:
Update NEMS commit

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   NEMS

Refs: #23

XianwuXue-NOAA committed:
Del fast-transcendentals to keep precise on Prod

And also update the GSDCHEM and WW3 to remove
	fast-transcendentals

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   GSDCHEM
	modified:   WW3
	modified:   conf/configure.fv3.wcoss2

Refs: #23

XianwuXue-NOAA committed:
Update the WW3 to new commit with deleting fast

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   WW3

Refs: #23

XianwuXue-NOAA committed:
Update GSDCHEM commit id and change to EMC repo

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules
	modified:   GSDCHEM

Refs: #23

XianwuXue-NOAA committed:
Change GSDCHEM to EMC repo and delete FV3 branch

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules

Refs: #23

WalterKolczynski-NOAA commented on .gitmodules:
```diff
@@ -1,12 +1,12 @@
 [submodule "GSDCHEM"]
 	path = GSDCHEM
-	url = gerrit:GSDCHEM
+	url = https://github.com/NOAA-EMC/GSDCHEM.git
```
Why does GSDCHEM have a .git extension and none of the others do?

XianwuXue-NOAA replied:
Thanks. I will modify it later

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @KateFriedman-NOAA @JessicaMeixner-NOAA Could you review my changes? One question: should I also modify modulefiles for hera and wcoss to make it build on these two machines? Because I changed "module load $(CONFDIR)/modules.nems" to ". $(CONFDIR)/modules.nems" to make XX_ver use default value or passing value from build.ver.

WalterKolczynski-NOAA commented on conf/configure.fv3.wcoss2:
```diff
@@ -0,0 +1,164 @@
+ifdef InNemsMakefile
+include $(TOP)/conf/configure.nems.NUOPC
```
Why does this file not have any indentation?

XianwuXue-NOAA replied:
I copied it from configure.fv3.wcoss_dell_p3 and  configure.fv3.wcoss_cray, to be consistent with the others, I didn't change the format. However, I am OK to reformat it.

WalterKolczynski-NOAA commented:
As long as the Hera module still works, I'll leave whether or not to update it to you. This repo is a dead-end anyway.

I removed Kate from the reviewer list, as she isn't appropriate for this.

WalterKolczynski-NOAA commented:
> I changed "module load $(CONFDIR)/modules.nems" to ". $(CONFDIR)/modules.nems" to make XX_ver use default value or passing value from build.ver.

If you change it this way, you are no longer loading a module, you are sourcing a shell script. You'd hardly be the first, but these are supposed to be modulefiles.

XianwuXue-NOAA commented:
Changing that way is to using default value of XX_ver or accept values from outside (like accept value from build.ver based on WCOSS2 new standard: "The build.ver file must be sourced before compiling the code"

XianwuXue-NOAA committed:
Correct typo

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   .gitmodules

Refs: #23

WalterKolczynski-NOAA commented:
The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)

XianwuXue-NOAA committed:
Delete the unused words in some statements

 On branch feature/gefs_v12_github
 Changes to be committed:
	modified:   conf/configure.fv3.wcoss2

Refs: #23

XianwuXue-NOAA commented:
> The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)

This way is also using on GFSv16  in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.

WalterKolczynski-NOAA commented:
> > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> 
> This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.

Rahul is already pushing to have those fixed too.

XianwuXue-NOAA commented:
> > > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> > 
> > 
> > This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.
> 
> Rahul is already pushing to have those fixed too.

Could you send me the link? I'd like to see Rahul's fixes.

WalterKolczynski-NOAA commented:
> > > > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> > > 
> > > 
> > > This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.
> > 
> > 
> > Rahul is already pushing to have those fixed too.
> 
> Could you send me the link? I'd like to see Rahul's fixes.

I don't think he's made any himself (yet), he's just trying to get them done.

XianwuXue-NOAA commented:
> > > > > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> > > > 
> > > > 
> > > > This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.
> > > 
> > > 
> > > Rahul is already pushing to have those fixed too.
> > 
> > 
> > Could you send me the link? I'd like to see Rahul's fixes.
> 
> I don't think he's made any himself (yet), he's just trying to get them done.

OK. Thanks for letting me know.  I am frozen the code for the canned data test. If he finds another way within one or two weeks, I can follow his way, otherwise, I think it is OK to keep this way. I have tested it works. Kate and Wen also confirmed.

WalterKolczynski-NOAA reviewed (approved):
Approved with the caveat we might turn modulefiles back into modulefiles instead of disguised shell scripts.

WalterKolczynski-NOAA commented:
Here's one Rahul did do. It is in lua instead of tcl: https://github.com/NOAA-EMC/obsproc/blob/develop/modulefiles/obsproc_wcoss2.lua

XianwuXue-NOAA commented:
> Here's one Rahul did do. It is in lua instead of tcl: https://github.com/NOAA-EMC/obsproc/blob/develop/modulefiles/obsproc_wcoss2.lua

Great. I checked the link and found it use "obsproc_common.lua" to overwrite the modules. I will see whether Kate and Wen will use this way in GFSv16 or not. If they use this way, I will try this way to unify with the global-workflow.

JessicaMeixner-NOAA reviewed (approved):
WW3 code looks good

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @JessicaMeixner-NOAA Thanks a lot for your review. If everything is OK, could you please merge this PR since I am not authorized to merge?

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA Could you please create a tag "gefs_v12.2.0"?

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA Could you please create a tag "gefs_v12.2.0"?

Are you sure this is ready? I'd prefer to wait until you run everything from this hash and make sure it is finalized.

XianwuXue-NOAA commented:
OK. I will do a test next week.

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA I have tested and results are identical. So could you create a tag "gefs_v12.2.0"? Thanks.
